### PR TITLE
Sn9 15 implement sample unpacking

### PR DESCRIPTION
**Pull request by cryptal-mc**

### CODE REVIEW

File: constants/__init__.py [hunk 1]
```diff
@@ -161,6 +162,27 @@
             ),
         ],
     ),
+    (
+        3_601_190,
```
Comment (Sid-Data-Universe) [inline constants/__init__.py:166]: Can we make a constant for this and use across the 3 places it is used now?

1) for the epsilon experiment, 1 for the competition start, 1 for the sample unpacking block.

File: neurons/validator.py [hunk 2]
```diff
@@ -716,14 +716,29 @@ async def run_step(self):
             competition.constraints, cache_dir=self.config.model_dir
         )
 
+        if cur_block > constants.sample_unpack_block:
```
Comment (Sid-Data-Universe) [inline neurons/validator.py:719]: technically should be >=

File: neurons/validator.py [hunk 3]
```diff
@@ -772,17 +787,20 @@ async def run_step(self):
...
                                 batches,
                                 self.config.device,
                                 tokenizer.eos_token_id,
+                                pack_samples
                             ),
-                            ttl=360,
+                            ttl=400,
                             mode="spawn",
                         )
+                        bt.logging.debug(f'Finished computing loss in {time.time() - start_loss} seconds')
```
Comment (Sid-Data-Universe) [inline neurons/validator.py:803]: We already use the sample which records this information right? Is it necessary to add the extra log here?

File: neurons/validator.py [hunk 4]
```diff
@@ -972,16 +997,18 @@ def log_step(
         table.add_column("win_rate", style="magenta")
         table.add_column("win_total", style="magenta")
         table.add_column("weights", style="magenta")
+        table.add_column("competition weights", style="magenta")
```
Comment (Sid-Data-Universe) [inline neurons/validator.py:1000]: competition_weights for uniformity

File: neurons/validator.py [hunk 5]
```diff
@@ -956,6 +977,10 @@ def log_step(
             "uids": uids,
             "uid_data": {},
         }
+
+        # The sub-competition weights
+        sub_competition_weights = torch.softmax(model_weights / constants.temperature, dim=0)
+
         for uid in uids:
             step_log["uid_data"][str(uid)] = {
                 "uid": uid,
```
Comment (Sid-Data-Universe) [inline neurons/validator.py:986]: Did you need to add the competition_weight to the step log as well for the leaderboards?

File: neurons/validator.py [hunk 6]
```diff
@@ -972,16 +997,18 @@ def log_step(
...
         table.add_column("competition", style="magenta")
-        for uid in uids:
+        for idx, uid in enumerate(uids):
             try:
                 table.add_row(
                     str(uid),
                     str(round(step_log["uid_data"][str(uid)]["average_loss"], 4)),
                     str(round(step_log["uid_data"][str(uid)]["win_rate"], 4)),
                     str(step_log["uid_data"][str(uid)]["win_total"]),
                     str(round(self.weights[uid].item(), 4)),
```
Comment (Sid-Data-Universe) [inline neurons/validator.py:1010]: do weights and sub competition weights not get indexed into the same way? the uid should be fine without the enumerate?

File: pretrain/validation.py [hunk 7]
```diff
@@ -179,13 +183,26 @@ def compute_losses(
...
+                    # For this reason, we want to ignore all but the
+                    # first EOS token (the real one)
+                    pad_mask = shift_labels == pad_token_id
+                    zeros = torch.zeros_like(shift_labels[...,:1])
+                    pad_mask = torch.cat((zeros, pad_mask[...,:-1]), dim=-1).bool()
+                    shift_labels[pad_mask] = -100
+
                 # Flatten the tokens
-                loss_fct = torch.nn.CrossEntropyLoss()
+                loss_fct = torch.nn.CrossEntropyLoss(ignore_index=-100) # -100 is actually the default value
```
Comment (Sid-Data-Universe) [inline pretrain/validation.py:205]: If it is the default value did you need to add it explicitly for some reason?

File: pretrain/dataset.py [hunk 8]
```diff
@@ -15,42 +15,156 @@
...
+            self.padded_buffer += [self.tokenizer.eos_token_id] * self._get_pad_size(input_ids=input_ids[:-1])
+
+    def __iter__(self):
+        self.buffer = self.used_buffer + self.buffer
+        self.padded_buffer = []
+
+        # Pad and prepare one page for batching
+        self._refill_padded_buffer()
+
+        return self
```
Comment (Sid-Data-Universe) [inline pretrain/dataset.py:129]: Previously this yielded the stack? This isn't needed anymore?

File: pretrain/dataset.py [hunk 9]
```diff
@@ -15,42 +15,156 @@
...
+    def __next__(self):
+        batch = []
+
+        while len(self.padded_buffer) >= self.sequence_length:
+            batch.append(self.padded_buffer[: self.sequence_length])
+            self.padded_buffer = self.padded_buffer[self.sequence_length :]
+            self._refill_padded_buffer()
+
+            if len(batch) == self.batch_size:
+                return np.stack(batch)
```
Comment (Sid-Data-Universe) [inline pretrain/dataset.py:140]: If we are just making it a torch.tensor in the compute_losses function do we need to make this a stack here?

Just trying to understand where the numpy is coming in.

File: pretrain/validation.py [hunk 10]
```diff
@@ -144,7 +144,11 @@ def check_for_reasonable_output(
 
 
 def compute_losses(
-    model, batches: typing.List[torch.Tensor], device: str, pad_token_id: int
+        model,
+        batches: typing.List[torch.Tensor],
+        device: str,
+        pad_token_id: int,
+        pack_samples: bool,
```
Comment (Sid-Data-Universe) [inline pretrain/validation.py:151]: This is not a bool to say if the samples should be packed, but saying if they were right?

Perhaps change parameter name to better indicate that. "packed_samples" or "sample_packing_used"

File: pretrain/validation.py [hunk 11]
```diff
@@ -144,7 +144,11 @@ def check_for_reasonable_output(
 
 
 def compute_losses(
-    model, batches: typing.List[torch.Tensor], device: str, pad_token_id: int
+        model,
+        batches: typing.List[torch.Tensor],
```
Comment (Sid-Data-Universe) [inline pretrain/validation.py:148]: As written this is getting a list of of numpy ndarrays? which you then convert into torch.tensors?

File: neurons/validator.py (hunk 2 above)
Comment (cryptal-mc) [reply neurons/validator.py:719]: Agree, done

File: neurons/validator.py (hunk 3 above)
Comment (cryptal-mc) [reply neurons/validator.py:803]: I am not sure where this is logged. I can however remove it, I added it while I was debugging.

File: neurons/validator.py (hunk 4 above)
Comment (cryptal-mc) [reply neurons/validator.py:1000]: done

File: pretrain/validation.py (hunk 7 above)
Comment (cryptal-mc) [reply pretrain/validation.py:205]: I added it that for clarity as some might not be aware of -100 as the default value. So setting EOS labels in the lines just before to -100 is more understandable. I will remove it, and keep the comment for clarity

File: pretrain/dataset.py (hunk 8 above)
Comment (cryptal-mc) [reply pretrain/dataset.py:129]: It is not. Actually, the __next__ function is the one that gets called. __iter__() can be used only to reset the iterator if we need to iterate again. This is a standard implementation also for __iter__() returning `self` and `__next__()` handling the yield logic (without using yield)

File: pretrain/dataset.py (hunk 9 above)
Comment (cryptal-mc) [reply pretrain/dataset.py:140]: Oh, we absolutely dont', we can just call torch.stack() in the `compute_losses` function. Let me actually do this, this could save us from importing numpy here

File: pretrain/validation.py (hunk 10 above)
Comment (cryptal-mc) [reply pretrain/validation.py:151]: that would be more accurate. I agree.
Done

File: pretrain/validation.py (hunk 11 above)
Comment (cryptal-mc) [reply pretrain/validation.py:148]: It is a list of list after I removed unnecessary numpy stacking

File: neurons/validator.py (hunk 2 above)
Comment (cryptal-mc) [reply neurons/validator.py:719]: Done

File: neurons/validator.py (hunk 6 above)
Comment (cryptal-mc) [reply neurons/validator.py:1010]: That's right
Done

File: pretrain/dataset.py (hunk 9 above)
Comment (cryptal-mc) [reply pretrain/dataset.py:140]: if we use torch, we first need to convert single entries in the batch to torch tensor

```
batch.append(torch.tensor(self.padded_buffer[: self.sequence_length]))
```

Then we stack those tensors at the end

```
torch.stack(batch)
```

Torch cannot stack the list of lists directly. Numpy can.

so if we return the list of lists, then in compute loss, we will need to iterate through each element in the batch to convert it to torch before stacking. I see using numpy is more convenient

File: neurons/validator.py [hunk 12]
```diff
@@ -965,23 +990,26 @@ def log_step(
                 "win_rate": win_rate[uid],
                 "win_total": wins[uid],
                 "weight": self.weights[uid].item(),
+                "norm_weight": sub_competition_weights[idx].item(),
```
Comment (RusticLuftig) [inline neurons/validator.py:993]: Optional: I find this name rather confusing. These are more "competition_weights", as you use in the table below. Both "weights" and these competition weights have been normalized, to some degree

File: neurons/validator.py [hunk 13]
```diff
@@ -1039,6 +1067,7 @@ def log_step(
                     str(uid): uid_data[str(uid)]["win_total"] for uid in uids
                 },
                 "weight_data": {str(uid): self.weights[uid].item() for uid in uids},
+                "norm_weight_data": {str(uid): sub_competition_weights[i].item() for i, uid in enumerate(uids)},
```
Comment (RusticLuftig) [inline neurons/validator.py:1070]: Same comment as above

**Review by RusticLuftig — APPROVED** [review]

File: neurons/validator.py (hunk 5 above)
Comment (RodrigoLPA) [reply neurons/validator.py:986]: I think so, it should help filtering the back end data.
